### PR TITLE
remove deadlock danger when removing Connections from a ConnectionMap

### LINKED ISSUE #1951: Avoid deadlock in Connection destructor
### Bug Report
When the TransportAdapterImpl removes a connection from the ConnectionMap, it will invoke the Connection destructor inside the "connections_lock_". This is undesirable, since it might cause a deadlock.

### PR DESCRIPTION
**Pull request by fronneburg**
Fixes #1951 

This PR is **[ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Summary
Created a new "RemoveConnection" method which safely removes a Connection object from the ConnectionMap.  Before calling "erase" on the connection object, the object's ref-count is incremented by assigning it to a local "ConnectionSPtr" variable. We then call "erase", NOT triggering the destructor, then release the "connections_lock_", and then let the local variable go out of scope, thus delaying the destructor until after the lock was released.

### CLA
- [X ] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc [hunk 1]
```diff
@@ -756,25 +748,53 @@ void TransportAdapterImpl::RemoveFinalizedConnection(
     const DeviceUID& device_handle, const ApplicationHandle& app_handle) {
   const DeviceUID device_uid = device_handle;
   LOG4CXX_AUTO_TRACE(logger_);
-  sync_primitives::AutoWriteLock lock(connections_lock_);
+  connections_lock_.AcquireForWriting();
```
Comment (LuxoftAKutsan) [inline src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc]: @fronneburg why did you remove AutoLock in that case?

File: src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc [hunk 2]
```diff
@@ -756,25 +748,53 @@ void TransportAdapterImpl::RemoveFinalizedConnection(
...
+  ConnectionSPtr connection = info.connection;
   connections_.erase(it_conn);
+  connections_lock_.Release();
+  LOG4CXX_TRACE(logger_, "RemoveFinalizedConnection Connections Lock Released");
 }
 
+void TransportAdapterImpl::RemoveConnection(
+    const DeviceUID& device_id, const ApplicationHandle& app_handle) {
+  ConnectionSPtr connection;
+  connections_lock_.AcquireForWriting();
```
Comment (LuxoftAKutsan) [inline src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc]: @fronneburg I would prefer using RAII here

File: src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc (hunk 1 above)
Comment (fronneburg) [reply src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc]: @LuxoftAKutsan we need to make sure that the lock is released first, and THEN let "connection" go out of scope.

File: src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc (hunk 2 above)
Comment (fronneburg) [reply src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc]: Same response here.

**Comment by JackLivio** [comment]
@fronneburg Why is it important that the calling the connection destructor is delayed until after the lock is released?

**Comment by fronneburg** [comment]
@JackLivio, we found a deadlock when the connection destructor (actually the "ThreadedConnection" Xevo implementation) tried to "join" a thread that was waiting on the "connections_lock_" in TransportAdapterImpl::ConnectionFinished.

**Comment by iCollin** [comment]
Closed in favor of https://github.com/smartdevicelink/sdl_core/pull/3371 which is based on latest develop and changes the log type from TRACE to DEBUG.
